File: server/app/services/weather_service.py
"""
Weather data service.
Fetches cloud cover from Open-Meteo API with caching.
"""

import logging
import time

import httpx

from app.utils.constants import OPEN_METEO_BASE, WEATHER_CACHE_S

logger = logging.getLogger(__name__)

# Cache keyed by grid cell (0.5° resolution)
_weather_cache: dict = {}


def _grid_key(lat: float, lon: float) -> str:
    """Round coordinates to 0.5° grid cell."""
    g_lat = f"{round(lat * 2) / 2:.1f}"
    g_lon = f"{round(lon * 2) / 2:.1f}"
    return f"{g_lat},{g_lon}"


async def fetch_cloud_cover(lat: float, lon: float) -> dict:
    """
    Fetch cloud cover for a given latitude and longitude.
    Returns {"total", "low", "mid", "high"} cloud cover percentages.
    Caches per 0.5° grid cell for 15 minutes.
    """
    key = _grid_key(lat, lon)
    now = time.time()

    # Check cache
    if key in _weather_cache:
        cached = _weather_cache[key]
        if now - cached["timestamp"] < WEATHER_CACHE_S:
            return cached["data"]

    try:
        url = f"{OPEN_METEO_BASE}?latitude={lat}&longitude={lon}&current=cloud_cover,cloud_cover_low,cloud_cover_mid,cloud_cover_high"
        async with httpx.AsyncClient() as client:
            resp = await client.get(url, timeout=10.0)
            resp.raise_for_status()
            json_data = resp.json()

        current = json_data.get("current", {})

        data = {
            "total": current.get("cloud_cover"),
            "low": current.get("cloud_cover_low"),
            "mid": current.get("cloud_cover_mid"),
            "high": current.get("cloud_cover_high"),
        }

        _weather_cache[key] = {"data": data, "timestamp": now}
        return data
    except Exception as exc:
        logger.error("fetch_cloud_cover error: %s", exc)
        # Return cached data if available (even if stale)
        if key in _weather_cache:
            return _weather_cache[key]["data"]
        return {"total": None, "low": None, "mid": None, "high": None}


def clear_cache():
    """Clear the weather cache (for testing)."""
    _weather_cache.clear()

File: server/src/services/weatherService.js
/**
 * Weather data service.
 * Fetches cloud cover from Open-Meteo API with caching.
 */
const fetch = require('node-fetch');
const { OPEN_METEO_BASE, WEATHER_CACHE_MS } = require('../utils/constants');

// Cache keyed by grid cell (0.5° resolution)
const weatherCache = new Map();

/**
 * Round coordinates to 0.5° grid cell.
 */
function gridKey(lat, lon) {
  const gLat = (Math.round(lat * 2) / 2).toFixed(1);
  const gLon = (Math.round(lon * 2) / 2).toFixed(1);
  return `${gLat},${gLon}`;
}

/**
 * Fetch cloud cover for a given latitude and longitude.
 * Returns { total, low, mid, high } cloud cover percentages.
 * Caches per 0.5° grid cell for 15 minutes.
 */
async function fetchCloudCover(lat, lon) {
  const key = gridKey(lat, lon);
  const now = Date.now();

  // Check cache
  if (weatherCache.has(key)) {
    const cached = weatherCache.get(key);
    if (now - cached.timestamp < WEATHER_CACHE_MS) {
      return cached.data;
    }
  }

  try {
    const url = `${OPEN_METEO_BASE}?latitude=${lat}&longitude=${lon}&current=cloud_cover,cloud_cover_low,cloud_cover_mid,cloud_cover_high`;
    const res = await fetch(url, { timeout: 10000 });
    if (!res.ok) throw new Error(`HTTP ${res.status} from Open-Meteo`);

    const json = await res.json();
    const current = json.current || {};

    const data = {
      total: current.cloud_cover != null ? current.cloud_cover : null,
      low: current.cloud_cover_low != null ? current.cloud_cover_low : null,
      mid: current.cloud_cover_mid != null ? current.cloud_cover_mid : null,
      high: current.cloud_cover_high != null ? current.cloud_cover_high : null,
    };

    weatherCache.set(key, { data, timestamp: now });
    return data;
  } catch (err) {
    console.error('[weatherService] fetchCloudCover error:', err.message);
    // Return cached data if available (even if stale)
    if (weatherCache.has(key)) {
      return weatherCache.get(key).data;
    }
    return { total: null, low: null, mid: null, high: null };
  }
}

/**
 * Clear the weather cache (for testing).
 */
function clearCache() {
  weatherCache.clear();
}

module.exports = {
  fetchCloudCover,
  clearCache,
  _gridKey: gridKey,
  _weatherCache: weatherCache,
};
